### ISSUE TITLE
arm64 ABI issues on apple silicon

### ISSUE THREAD

**Issue by hdooley** [odin-lang/Odin#1272]
## Context

	Odin: dev-2021-11:887ae0ab
	OS:   macOS 12.0.0 Monterey
	CPU:  ARM64
	RAM:  32768 MiB

## Expected Behavior

compiling the following program should result in a working executable:

```
package bad_allocator_init

import "core:log"

logger: log.Logger = log.create_console_logger(ident = "foo");

main :: proc() {
  context.logger = logger;
}

```

## Current Behavior

```
$ ./bad_allocator_init
zsh: bus error  ./bad_allocator_init
```

## Failure Information (for bugs)

The LLVM IR looks valid, but in the debugger i noticed what looked like a register mismatch in this function:

```
heap_allocator_proc :: proc(allocator_data: rawptr, mode: mem.Allocator_Mode,
                            size, alignment: int,
                            old_memory: rawptr, old_size: int, loc := #caller_location) -> ([]byte, mem.Allocator_Error) {

        // BUG HERE: Size should be 24 and align should be eight Instead, Size is 0 and alignment is 24.
	aligned_alloc :: proc(size, alignment: int, old_ptr: rawptr = nil) -> ([]byte, mem.Allocator_Error) {
		a := max(alignment, align_of(rawptr))
		space := size + a - 1

		allocated_mem: rawptr
		if old_ptr != nil {
			original_old_ptr := mem.ptr_offset((^rawptr)(old_ptr), -1)^
			allocated_mem = heap_resize(original_old_ptr, space+size_of(rawptr))
		} else {
			allocated_mem = heap_alloc(space+size_of(rawptr))
		}
		aligned_mem := rawptr(mem.ptr_offset((^u8)(allocated_mem), size_of(rawptr)))

		ptr := uintptr(aligned_mem)
		aligned_ptr := (ptr - 1 + uintptr(a)) & -uintptr(a)
		diff := int(aligned_ptr - ptr)
		if (size + diff) > space {
			return nil, .Out_Of_Memory
		}

		aligned_mem = rawptr(aligned_ptr)
		mem.ptr_offset((^rawptr)(aligned_mem), -1)^ = allocated_mem

		return mem.byte_slice(aligned_mem, size), nil
	}
```

[bad_allocator_init intermediates.zip](https://github.com/odin-lang/Odin/files/7480075/bad_allocator_init.intermediates.zip)

**Comment by hdooley** [comment]
For fun, I tried removing some inlining and moving aligned_alloc out into file scope but predictably this had no effect.

**Comment by hdooley** [comment]
Did some investigation with @Platin21 today, and here's what we discovered:

Though the issue above manifested as a bus_error, this was actually a failed assertion inside of os_darwin's heap_malloc procedure. Size was being passed as 0 due to the weird register mapping. replacing the call to _unix_calloc to just return nil yielded code that no longer failed that assertion, and the size parameter passed to heap_alloc appeared to be correct. Example of the change: 

```
heap_alloc :: proc(size: int) -> rawptr {
	assert(size > 0);
	hax : uintptr = 0xDEADBEEF;
	return transmute(rawptr)hax;
}
```

Replacing `_unix_calloc` with `_unix_malloc` + memset to zero didn't change the situation.

On a hunch, Platin suggested calling unix_malloc directly in main to see if we could get memory from the OS at all. To my surprise this actually worked. Example:

```
STDOUT :: os.Handle(0)

console_raw_write :: proc(str: string) {
  slice := transmute([]u8)str
  os.write(STDOUT, slice)
}

main :: proc() {

  console_raw_write("Malloc for memory\n");
  ptr : [^]u8 = transmute([^]u8)os._unix_malloc(size_of(u8) * 30);


  for i := 0; i < 30; i += 1 {
    ptr[i] = u8(i) + 65
  }

  for i := 0; i < 30; i += 1 {
    os.write_byte(STDOUT, ptr[i])
  }
}
```

This produced the expected output.

One last idea we tested was using static arena allocator with an explicitly initialized default context. This produced a crash similar to the initial issue. Example code:

```
arena_buf : [1024 * 1024]u8;

test_arena_allocator :: proc "contextless" () -> []u8 {

  context = runtime.default_context();

  arena : mem.Arena;
  mem.init_arena(&arena, arena_buf[:]);

  allocator := mem.arena_allocator(&arena);

  foo := make([]u8, 30, allocator);

  return foo;
}

main :: proc() {
  console_raw_write("Static Arena for memory\n");

  static_ptr := test_arena_allocator(); // bus_error here

  for i := 0; i < 30; i += 1 {
    static_ptr[i] = u8(i) + 65
  }

  for i := 0; i < 30; i += 1 {
    os.write_byte(STDOUT, static_ptr[i])
  }
}
```

Not exactly sure where this leaves us in terms of root-cause, but my initial guess that it was somehow related to making a foreign call appears to be incorrect. It seems related somehow to odin's context system, but it's difficult to say for sure what the problem is. @Platin21 noticed some strangeness around aggregated return types in the llvm-ir, but manually fixing the IR and assembling didn't appear to fix the issue.

**Comment by hdooley** [comment]
Here's the full example if anyone wants to look into it further.
```
package main

import "core:os"
import "core:fmt"
import "core:strings"
import "core:mem"
import "core:runtime"

STDOUT :: os.Handle(0)

console_raw_write :: proc(str: string) {
  slice := transmute([]u8)str
  os.write(STDOUT, slice)
}

arena_buf : [1024 * 1024]u8;

test_arena_allocator :: proc "contextless" () -> []u8 {

  context = runtime.default_context();

  arena : mem.Arena;
  mem.init_arena(&arena, arena_buf[:]);

  allocator := mem.arena_allocator(&arena);

  foo := make([]u8, 30, allocator);

  return foo;
}

main :: proc() {

  console_raw_write("Malloc for memory\n");
  ptr : [^]u8 = transmute([^]u8)os._unix_malloc(size_of(u8) * 30);

  for i := 0; i < 30; i += 1 {
    ptr[i] = u8(i) + 65
  }

  for i := 0; i < 30; i += 1 {
    os.write_byte(STDOUT, ptr[i])
  }
  
  console_raw_write("\nStatic Arena for memory\n");

  static_ptr := test_arena_allocator();

  for i := 0; i < 30; i += 1 {
    static_ptr[i] = u8(i) + 65
  }

  for i := 0; i < 30; i += 1 {
    os.write_byte(STDOUT, static_ptr[i])
  }
}

```

Output:
```
$ ./odin build bad_alloc.odin -debug -opt:0
$ ./bad_alloc
Malloc for memory
ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^
Static Arena for memory
zsh: bus error  ./bad_alloc
```

**Comment by hdooley** [comment]
As a test, i hacked together a contextless arena allocator to narrow down the issue further. This sample also bus_errors inside of the allocation procedure once it performs the indirect call through the proc pointer on the allocator struct.

```
package main

import "core:os"
import "core:fmt"
import "core:strings"
import "core:mem"
import "core:runtime"

STDOUT :: os.Handle(0)

console_raw_write :: proc(str: string) {
  slice := transmute([]u8)str
  os.write(STDOUT, slice)
}

HACK_DEFAULT_ALIGNMENT :: 2*align_of(rawptr)

HackAllocator_Error :: enum byte {
  None = 0,
  Out_Of_Memory,
}

HackAllocator_Mode :: enum byte {
  Alloc,
  Free,
  Free_All,
  Resize,
  Query_Features,
}

HackAllocator_Proc :: #type proc "contextless" (allocator_data: rawptr, mode: HackAllocator_Mode,
                             size, alignment: int,
                             old_memory: rawptr, old_size: int,
                             location: runtime.Source_Code_Location = #caller_location) -> ([]byte, HackAllocator_Error);
HackAllocator :: struct {
  procedure: HackAllocator_Proc,
  data:      rawptr,
}

HackArena :: struct {
  data   : []byte,
  offset : uintptr,
}

hack_arena : HackArena;

HackArenaAlloc :: proc "contextless" (allocator_data: rawptr, mode: HackAllocator_Mode,
                       size, alignment: int,
                       old_memory: rawptr, old_size: int,
                       location: runtime.Source_Code_Location = #caller_location) -> ([]byte, HackAllocator_Error) {
  arena := transmute(^HackArena)allocator_data;
  // assert(uintptr(len(arena.data)) - arena.offset >= uintptr(size));

  offset : uintptr = uintptr((transmute(mem.Raw_Slice)arena.data).data) + arena.offset;
  bytes := transmute([]byte)mem.Raw_Slice{rawptr(offset), size};
  arena.offset += uintptr(size);

  return bytes, .None;
}

hack_alloc_bytes :: proc "contextless" (size: int, alignment: int, allocator: HackAllocator, loc := #caller_location) -> ([]byte, HackAllocator_Error) {
  if size == 0 {
    return nil, nil
  }
  if allocator.procedure == nil {
    return nil, nil
  }
  return allocator.procedure(allocator.data, HackAllocator_Mode.Alloc, size, alignment, nil, 0, loc)
}

hack_make_aligned :: proc "contextless" ($T: typeid/[]$E, #any_int len: int, alignment: int, allocator: HackAllocator, loc := #caller_location) -> (slice: T, err: HackAllocator_Error) {
  data := hack_alloc_bytes(size_of(E)*len, alignment, allocator, loc) or_return
  if data == nil && size_of(E) != 0 {
    return
  }
  slice = transmute(T)data;
  return
}

hack_make_slice :: proc "contextless" ($T: typeid/[]$E, #any_int len: int, allocator: HackAllocator, loc := #caller_location) -> (T, HackAllocator_Error) {
  return hack_make_aligned(T, len, align_of(E), allocator, loc)
}

hack_make :: proc {
  hack_make_slice,
}


arena_buf : [1024 * 1024]u8;

test_arena_allocator :: proc "contextless" () -> []u8 {

  // context = runtime.default_context();

  hack_arena.data = arena_buf[:];

  allocator : HackAllocator;
  allocator.procedure = HackArenaAlloc;
  allocator.data = &hack_arena;

  foo, err := hack_make([]u8, 30, allocator);

  return foo;
}

main :: proc() {

  console_raw_write("Malloc for memory\n");
  ptr : [^]u8 = transmute([^]u8)os._unix_malloc(size_of(u8) * 30);

  for i := 0; i < 30; i += 1 {
    ptr[i] = u8(i) + 65
  }

  for i := 0; i < 30; i += 1 {
    os.write_byte(STDOUT, ptr[i])
  }
  
  console_raw_write("\nStatic Arena for memory\n");

  static_ptr := test_arena_allocator();

  for i := 0; i < 30; i += 1 {
    static_ptr[i] = u8(i) + 65
  }

  for i := 0; i < 30; i += 1 {
    os.write_byte(STDOUT, static_ptr[i])
  }
}
```
```
$ ./odin build bad_alloc.odin -debug -opt:0
$ ./bad_alloc
Malloc for memory
ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^
Static Arena for memory
zsh: bus error  ./bad_alloc
```

The cause of the bus_error is that the HackArena is invalid inside of that call, returning zero instead of a valid address.
Changing the allocator to instead call the procedure directly results in a working program:

```
HackArenaAlloc :: proc "contextless" (allocator_data: rawptr, mode: HackAllocator_Mode,
                       size, alignment: int,
                       old_memory: rawptr, old_size: int,
                       location: runtime.Source_Code_Location = #caller_location) -> ([]byte, HackAllocator_Error) {
  arena := transmute(^HackArena)allocator_data;
  // assert(uintptr(len(arena.data)) - arena.offset >= uintptr(size));

  offset : uintptr = uintptr((transmute(mem.Raw_Slice)arena.data).data) + arena.offset;
  bytes := transmute([]byte)mem.Raw_Slice{rawptr(offset), size};
  arena.offset += uintptr(size);

  return bytes, .None;
}

hack_alloc_bytes :: proc "contextless" (size: int, alignment: int, allocator: HackAllocator, loc := #caller_location) -> ([]byte, HackAllocator_Error) {
  if size == 0 {
    return nil, nil
  }
  if allocator.procedure == nil {
    return nil, nil
  }
  // THIS WORKS
  return HackArenaAlloc(allocator.data, HackAllocator_Mode.Alloc, size, alignment, nil, 0, loc)

  // THIS DOESNT
  // return allocator.procedure(allocator.data, HackAllocator_Mode.Alloc, size, alignment, nil, 0, loc)
}
```

```
$ ./odin build bad_alloc.odin -debug -opt:0
$ ./bad_alloc
Malloc for memory
ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^
Static Arena for memory
ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^%
```

My conclusion is that the issue appears to be with indirect calls. I'll get a minimal example put together, but we should look into what's going on in the llvm backend as well.

**Comment by Platin21** [comment]
This is fixed with #1386 can be closed
